distributed/shuffle/_arrow.py

```diff
@@ -85,3 +77,20 @@ def deserialize_table(buffer: bytes) -> pa.Table:
 
     with pa.ipc.open_stream(pa.py_buffer(buffer)) as reader:
         return reader.read_all()
+
+
+def read_from_disk(path: Path, meta: pd.DataFrame) -> tuple[Any, int]:
+    import pyarrow as pa
+
+    shards = []
+    schema = pa.Schema.from_pandas(meta, preserve_index=True)
+    with pa.OSFile(str(path), mode="rb") as f:
+        size = f.seek(0, whence=2)
+        f.seek(0)
+        while f.tell() < size:
+            sr = pa.RecordBatchStreamReader(f)
+            shard = sr.read_all()
+            arrs = [pa.concat_arrays(column.chunks) for column in shard.columns]
```

Comment:
From what I understand, the `RecordBatchStreamReader` creates one buffer per record batch. On `main`, this is a problem when we convert the `pa.Table` consisting of all those batches into a `pd.DataFrame`. This conversion frees buffers on a per-column basis. Effectively, this means that all buffers from all record batches will not be freed until we converted the last column. To avoid this, we force a copy for each column directly after reading it with `pa.concat_arrays`. This way, we (should) have one buffer per column per batch.

Reply:
Similarly, `pa.Table.combine_chunks` proceeds on a per-column basis causing a spike in temporary memory usage (see #8128).